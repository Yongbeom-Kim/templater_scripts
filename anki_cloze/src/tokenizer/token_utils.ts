export const isSpace = (char: string): boolean => {
  return char == " ";
};

export const isNewline = (char: string): boolean => {
  return char == "\n";
};
