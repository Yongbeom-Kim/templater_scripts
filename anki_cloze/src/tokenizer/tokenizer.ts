import { Token, TokenizerState, TokenType } from "./token_types";

// TODO: Consider converting to stream-based tokenizer if performance becomes an issue
export const tokenize = (text: string): Token[] => {
  // Normalize newlines
  text = text.replaceAll("\r\n", "\n").replaceAll("\r", "\n");

  const tokens: Token[] = [];
  let state = new TokenizerState(text);
  while (state.good()) {
    let lexeme;
    let updatedState;
    if (isText(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isText));
      tokens.push(Token.Create(state, TokenType.Text, lexeme));
    } else if (isDigit(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isDigit));
      tokens.push(Token.Create(state, TokenType.Number, lexeme));
    } else if (isSpace(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isSpace));
      tokens.push(Token.Create(state, TokenType.Whitespace, lexeme));
    } else if (isNewline(state.peek(1))) {
      [updatedState, lexeme] = state.consume(1);
      tokens.push(Token.Create(state, TokenType.Newline, lexeme));
    } else if (isPunctuation(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isPunctuation));
      tokens.push(Token.Create(state, TokenType.Punctuation, lexeme));
    } else {
      throw new Error(`Unexpected character: "${state.peek(1)}". Index: (${state.line_pos}, ${state.col_pos}). Context: ${state.peek(5)}...`);
    }
    state = updatedState;
  }
  return tokens;
};

const isText = (c: string): boolean => {
  return !!c.match(/[a-zA-Z_]/);
};

const isDigit = (c: string): boolean => {
  return !!c.match(/\d/);
};

const isSpace = (c: string): boolean => {
  // \u00A0 is a non-breaking space
  return c == " " || c == "\t" || c == "\u00A0";
};

const isNewline = (c: string): boolean => {
  return c == "\n";
};

const isPunctuation = (c: string): boolean => {
  return !!c.match(/[^\w\s]/);
};

const not = (predicate: (...args: any[]) => boolean) => {
  return (...args: any[]) => !predicate(...args);
};
