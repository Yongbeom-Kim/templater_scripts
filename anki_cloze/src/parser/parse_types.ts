import { Token, TokenType } from "../tokenizer/token_types";

export class ParserState {
  constructor(
    public readonly tokens: Token[],
    public readonly next: number = 0,
  ) {}

  good(): boolean {
    return this.next < this.tokens.length;
  }

  eof(): boolean {
    return this.next >= this.tokens.length;
  }

  isStartOfLine(): boolean {
    return (
      this.next == 0 || this.tokens[this.next - 1].type === TokenType.Newline
    );
  }

  peek(n_tokens: number = 1): Token[] {
    return this.tokens.slice(this.next, this.next + n_tokens);
  }

  consume(n_tokens: number = 1): [ParserState, Token[]] {
    const result = this.tokens.slice(this.next, this.next + n_tokens);
    return [
      new ParserState(
        this.tokens,
        Math.min(this.next + n_tokens, this.tokens.length),
      ),
      result,
    ];
  }

  consumeUntilType(
    type: TokenType,
    include: boolean = false,
  ): [ParserState, Token[]] {
    const result: Token[] = [];
    let i = this.next;
    for (; i < this.tokens.length && this.tokens[i].type !== type; i++) {
      result.push(this.tokens[i]);
    }
    if (include && i < this.tokens.length) {
      result.push(this.tokens[i]);
      i++;
    }
    return [new ParserState(this.tokens, i), result];
  }

  consumeOnlyType(type: TokenType): [ParserState, Token[]] {
    const result: Token[] = [];
    let i = this.next;
    for (; i < this.tokens.length && this.tokens[i].type === type; i++) {
      result.push(this.tokens[i]);
    }
    return [new ParserState(this.tokens, i), result];
  }

  consumeUntilLexeme(
    lexeme: string,
    include: boolean = false,
  ): [ParserState, Token[]] {
    const result: Token[] = [];
    let i = this.next;
    for (; i < this.tokens.length && this.tokens[i].lexeme !== lexeme; i++) {
      result.push(this.tokens[i]);
    }
    if (include && i < this.tokens.length) {
      result.push(this.tokens[i]);
      i++;
    }
    return [new ParserState(this.tokens, i), result];
  }

  debug(): string {
    if (this.eof()) {
      return `
      Parser State:
        Document: "${this.tokens.map((t) => t.lexeme).join("")}"
        Index: "${this.next}"
        Good: "${this.good()}"
        EOF: "${this.eof()}"
      `;
    }
    return `
    Parser State:
      Document: "${this.tokens.map((t) => t.lexeme).join("")}"
      Index: "${this.next}"
      Good: "${this.good()}"
      EOF: "${this.eof()}"
      Peek: ${this.peek()[0].lexeme}
      Peek_type: ${this.peek()[0].type}
      Peek (5): "${this.peek(5)
        .map((t) => t.lexeme)
        .join("|")}"
    `;
  }
}
export enum ParseTreeNodeType {
  Indent = "indent",
  Text = "text",
  List = "list",
  CodeBlock = "code_block",
  CodeLine = "code_line",
  CodeComment = "code_comment",
}

export abstract class ParseTreeVisitor {
  abstract visit(node: ParseTreeNode): void;
}

export abstract class ParseTreeNode {
  abstract type: ParseTreeNodeType;
  visit(visitor: ParseTreeVisitor) {
    visitor.visit(this);
  }
  abstract toText(): string;

  static toText(tree: ParseTreeNode[]): string {
    return tree.map((n) => n.toText()).join("");
  }
}

export class IndentNode extends ParseTreeNode {
  type = ParseTreeNodeType.Indent;
  n_spaces: number;
  n_tabs: number;
  private constructor(public readonly indent?: Token) {
    super();
    if (indent && indent.type !== TokenType.Whitespace) {
      throw new Error("Indent must be a whitespace token");
    }
    this.n_spaces = [...(indent?.lexeme ?? "")].filter((c) => c === " ").length;
    this.n_tabs = [...(indent?.lexeme ?? "")].filter((c) => c === "\t").length;
  }
  toText(): string {
    return this.indent?.lexeme ?? "";
  }

  static FromWhitespace(whitespace: Token): IndentNode {
    return new IndentNode(whitespace);
  }
}

/**
 * Node that contains some text.
 * This is the most basic node type.
 */
export class TextNode extends ParseTreeNode {
  type = ParseTreeNodeType.Text;
  constructor(public readonly contents: Token[]) {
    super();
  }
  toText(): string {
    return this.contents.map((t) => t.lexeme).join("");
  }

  static FromToken(token: Token): TextNode {
    return new TextNode([token]);
  }

  static FromTokens(tokens: Token[]): TextNode {
    return new TextNode(tokens);
  }
}

/**
 * A text line node is a node that contains a list of tokens that are not a list or a header.
 *
 * Text line nodes encompass one entire line of text,
 *  excluding the newline token at the start,
 *  including the newline token at the end.
 */
export class TextLineNode extends ParseTreeNode {
  type = ParseTreeNodeType.Text;

  constructor(
    public readonly indent: IndentNode,
    public readonly contents: ParseTreeNode[],
    public readonly endingNewline?: Token, // undefined if EOF
  ) {
    super();
  }
  toText(): string {
    return (
      this.indent.toText() +
      this.contents.map((t) => t.toText()).join("") +
      (this.endingNewline?.lexeme ?? "")
    );
  }
}

export class ListNode extends TextLineNode {
  type = ParseTreeNodeType.List;

  constructor(
    public readonly ordered: boolean,
    public readonly indent: IndentNode,
    public readonly marker: Token[],
    public readonly contents: ParseTreeNode[],
    public readonly endingNewline?: Token, // undefined if EOF
  ) {
    super(indent, contents, endingNewline); // This should combine indent + marker + contents, but it's okay since we don't rely on the inheritance.
  }
  toText(): string {
    return (
      this.indent.toText() +
      this.marker.map((t) => t.lexeme).join("") +
      " " +
      this.contents.map((t) => t.toText()).join("") +
      (this.endingNewline?.lexeme ?? "")
    );
  }
}

export enum CodeBlockLanguage {
  Css = "css",
  C = "c",
  Cpp = "cpp",
  Go = "go",
  Groovy = "groovy",
  Hcl = "hcl",
  JavaScript = "javascript",
  Jsx = "jsx",
  TypeScript = "typescript",
  Tsx = "tsx",
  Python = "python",
  Plsql = "plsql",
  Sql = "sql",
  Toml = "toml",
  Yaml = "yaml",
  None = "",
}

export namespace CodeBlockLanguage {
  // TODO: move to parser, this shouldn't be here
  export const FromKeyword = (keyword?: string): CodeBlockLanguage => {
    if (!keyword) {
      console.warn(
        "No keyword provided when parsing code block language. Expected a language name.",
      );
      return CodeBlockLanguage.None;
    }
    switch (keyword.toLowerCase()) {
      case "css":
        return CodeBlockLanguage.Css;
      case "c":
        return CodeBlockLanguage.C;
      case "cpp":
      case "c++":
        return CodeBlockLanguage.Cpp;
      case "go":
        return CodeBlockLanguage.Go;
      case "groovy":
        return CodeBlockLanguage.Groovy;
      case "hcl":
        return CodeBlockLanguage.Hcl;
      case "javascript":
      case "js":
        return CodeBlockLanguage.JavaScript;
      case "jsx":
        return CodeBlockLanguage.Jsx;
      case "typescript":
      case "ts":
        return CodeBlockLanguage.TypeScript;
      case "tsx":
        return CodeBlockLanguage.Tsx;
      case "python":
      case "py":
        return CodeBlockLanguage.Python;
      case "plsql":
        return CodeBlockLanguage.Plsql;
      case "sql":
        return CodeBlockLanguage.Sql;
      case "toml":
        return CodeBlockLanguage.Toml;
      case "yaml":
      case "yml":
        return CodeBlockLanguage.Yaml;
      default:
        console.warn(
          `Unrecognized keyword when parsing code block language: ${keyword}`,
        );
        return CodeBlockLanguage.None;
    }
  };
}

export class CodeBlockNode extends ParseTreeNode {
  type = ParseTreeNodeType.CodeBlock;
  constructor(
    public readonly language_str: string,
    public readonly language: CodeBlockLanguage,
    public readonly contents: CodeLineNode[],
    public readonly endingNewline?: Token, // undefined if EOF
  ) {
    super();
    for (const content of contents) {
      if (
        content.type !== ParseTreeNodeType.CodeLine &&
        content.type !== ParseTreeNodeType.CodeComment
      ) {
        throw new Error(
          "Code block contents must be code line or code comment nodes",
        );
      }
    }
  }

  toText(): string {
    return `\`\`\`${this.language_str}\n${this.contents
      .map((t) => t.toText())
      .join("")}\`\`\`${this.endingNewline?.lexeme ?? ""}`;
  }
}

export class CodeLineNode extends ParseTreeNode {
  type = ParseTreeNodeType.CodeLine;
  constructor(
    public readonly indent: IndentNode,
    public readonly contents: ParseTreeNode[],
    public readonly endingNewline?: Token, // undefined if EOF
  ) {
    super();
  }
  toText(): string {
    return (
      this.indent.toText() +
      this.contents.map((t) => t.toText()).join("") +
      (this.endingNewline?.lexeme ?? "")
    );
  }
  empty(): boolean {
    // TODO: Do better
    return this.toText().trim().length === 0;
  }
}

export class CodeCommentNode extends CodeLineNode {
  type = ParseTreeNodeType.CodeComment;
}
