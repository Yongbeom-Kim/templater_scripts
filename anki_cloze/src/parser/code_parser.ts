import { Token, TokenType } from "../tokenizer/token_types";
import { CodeBlockLanguage, ParserState } from "./parse_types";

export const isComment = (
  language: CodeBlockLanguage,
  state: ParserState,
): boolean => {
  if (state.eof()) {
    throw new Error(
      `Parser state in EOF in middle of code block. ${state.debug()}`,
    );
  }

  [state] = state.consumeOnlyType(TokenType.Whitespace);

  switch (language) {
    case CodeBlockLanguage.None:
      return state.peek()[0].lexeme === "#" || state.peek()[0].lexeme === "//";
    case CodeBlockLanguage.Python:
      return state.peek()[0].lexeme === "#";
    case CodeBlockLanguage.Cpp:
      return state.peek()[0].lexeme === "/*" || state.peek()[0].lexeme === "//";
    default:
      return false;
  }
};
