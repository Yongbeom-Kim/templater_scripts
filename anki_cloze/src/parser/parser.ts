import { Token, TokenType } from "../tokenizer/token_types";
import { isComment } from "./code_parser";
import {
  CodeBlockNode,
  CodeBlockLanguage,
  CodeCommentNode,
  CodeLineNode,
  IndentNode,
  ListNode,
  ParseTreeNode,
  ParserState,
  TextLineNode,
  TextNode,
} from "./parse_types";

export const parse = (tokens: Token[]): ParseTreeNode[] => {
  const result: ParseTreeNode[] = [];
  let state = new ParserState(tokens);
  while (state.good()) {
    let node: ParseTreeNode | null = null;
    [state, node] = tryParseCodeBlock(state);
    if (node) {
      result.push(node);
      continue;
    }
    [state, node] = tryParseLine(state);
    if (node) {
      result.push(node);
      continue;
    }
    throw new Error(
      `Unreachable code reached in parse function. Document: "${tokens
        .map((t) => t.lexeme)
        .join("")}", Token: "${state.peek()[0].lexeme}", Index: ${state.next}`,
    );
  }
  return result;
};

/**
 * Attempts to parse a code block from the current position
 */
const tryParseCodeBlock = (
  state: ParserState,
): [ParserState, ParseTreeNode | null] => {
  if (!state.isStartOfLine()) {
    return [state, null];
  }

  const backupState = state;
  let peeked: Token[] = [];
  let language: CodeBlockLanguage, language_str: string, backTick: Token;
  // parse starting backtick
  [state, peeked] = state.consumeUntilType(TokenType.Newline, true);
  if (
    !(
      peeked[0] &&
      peeked[0].type === TokenType.Punctuation &&
      peeked[0].lexeme === "```"
    )
  ) {
    return [backupState, null];
  }
  backTick = peeked[0];
  if (peeked.length == 2) {
    language = CodeBlockLanguage.None;
    language_str = "";
  } else {
    language_str = peeked
      .slice(1)
      .map((t) => t.lexeme)
      .join("")
      .trim();
    language = CodeBlockLanguage.FromKeyword(language_str);
  }

  // parse body
  const children: CodeLineNode[] = [];
  let parsed: CodeLineNode | null = null;
  while (state.good()) {
    [state, parsed] = parseCodeLine(language, state);
    children.push(parsed);
    if (
      state.peek()[0].type === TokenType.Punctuation &&
      state.peek()[0].lexeme === backTick.lexeme
    ) {
      break;
    }
  }

  // parse ending backtick
  [state, peeked] = state.consumeUntilType(TokenType.Newline, true);
  const endingBackTick = peeked[0];
  if (
    endingBackTick.type !== TokenType.Punctuation ||
    endingBackTick.lexeme !== "```"
  ) {
    return [backupState, null];
  }
  const endingNewline =
    peeked.length > 1 ? peeked[peeked.length - 1] : undefined;
  return [
    state,
    new CodeBlockNode(language_str, language, children, endingNewline),
  ];
};

const parseCodeLine = (
  language: CodeBlockLanguage,
  state: ParserState,
): [ParserState, CodeLineNode] => {
  let indent, contents, endingNewline, comment;
  [state, indent] = state.consumeOnlyType(TokenType.Whitespace);
  if (indent.length > 1) {
    throw new Error(
      `Assertion failed in parseCodeLine function. Whitespace tokens should all be combined into one token. Indent: "${indent.map((t) => t.lexeme).join("")}", Token Chain: "${state
        .peek()
        .map((t) => t.lexeme)
        .join("")}"`,
    );
  }
  comment = isComment(language, state);
  [state, contents] = state.consumeUntilType(TokenType.Newline, false);
  [state, [endingNewline]] = state.consume();
  if (comment) {
    return [
      state,
      new CodeCommentNode(
        IndentNode.FromWhitespace(indent[0]),
        [TextNode.FromTokens(contents)],
        endingNewline,
      ),
    ];
  }
  return [
    state,
    new CodeLineNode(
      IndentNode.FromWhitespace(indent[0]),
      [TextNode.FromTokens(contents)],
      endingNewline,
    ),
  ];
};

/**
 * Attempts to parse either a list item or a text line from the current position
 */
const tryParseLine = (
  state: ParserState,
): [ParserState, ParseTreeNode | null] => {
  if (!state.isStartOfLine()) {
    return [state, null];
  }

  let newState = state,
    indent,
    marker,
    content,
    endingNewline;
  [newState, indent] = state.consumeOnlyType(TokenType.Whitespace);
  if (indent.length > 1) {
    throw new Error(
      `Asssertion failed in tryParseLine function. Whitespace tokens should all be combined into one token. Indent: "${indent
        .map((t) => t.lexeme)
        .join("|")}", Token Chain: "${state
        .peek()
        .map((t) => t.lexeme)
        .join("")}"`,
    );
  }

  // Check for list markers first
  if (peekUnorderedListMarker(newState) || peekOrderedListMarker(newState)) {
    const ordered = peekOrderedListMarker(newState);
    [newState, marker] = newState.consumeUntilType(TokenType.Whitespace, false);
    [newState] = newState.consume();
    [newState, content] = newState.consumeUntilType(TokenType.Newline, false);
    [newState, [endingNewline]] = newState.consume();
    return [
      newState,
      new ListNode(
        ordered,
        IndentNode.FromWhitespace(indent[0]),
        marker,
        [TextNode.FromTokens(content)],
        endingNewline,
      ),
    ];
  }

  // If not a list, treat as regular text line
  [newState, content] = newState.consumeUntilType(TokenType.Newline, false);
  [newState, [endingNewline]] = newState.consume();
  if (content.length > 0 || endingNewline || newState.eof()) {
    return [
      newState,
      new TextLineNode(
        IndentNode.FromWhitespace(indent[0]),
        [TextNode.FromTokens(content)],
        endingNewline,
      ),
    ];
  }

  console.warn(
    `Start of line, but no tokens found. This should be unreachable. ${newState.debug()}`,
  );
  return [state, null];
};

const getIndentLevel = (tokens: Token[]): number => {
  const level = tokens.reduce((acc, token) => {
    if (token.type === TokenType.Whitespace) {
      for (let i = 0; i < token.lexeme.length; i++) {
        if (token.lexeme[i] === " ") {
          acc += 1 / 2;
        } else if (token.lexeme[i] === "\t") {
          acc += 1;
        } else {
          throw new Error(
            `Unreachable code reached in getIndentLevel function. Token: "${
              token.lexeme
            }", Token Chain: "${tokens.map((t) => t.lexeme).join("")}"`,
          );
        }
      }
      return acc;
    }
    throw new Error(
      `Unreachable code reached in getIndentLevel function. Token: "${
        token.lexeme
      }", Token Chain: "${tokens.map((t) => t.lexeme).join("")}"`,
    );
  }, 0);
  return Math.ceil(level);
};

const peekUnorderedListMarker = (state: ParserState): boolean => {
  const tokens = state.peek(2);
  return (
    tokens.length >= 2 &&
    tokens[0].type === TokenType.Punctuation &&
    (tokens[0].lexeme === "-" || tokens[0].lexeme === "*") &&
    tokens[1].type === TokenType.Whitespace
  );
};

const peekOrderedListMarker = (state: ParserState): boolean => {
  const tokens = state.peek(3);
  return (
    tokens.length >= 3 &&
    tokens[0].type === TokenType.Number &&
    tokens[1].type === TokenType.Punctuation &&
    tokens[1].lexeme === "." &&
    tokens[2].type === TokenType.Whitespace
  );
};
