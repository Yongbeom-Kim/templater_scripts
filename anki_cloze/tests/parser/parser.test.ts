import { describe, test, expect } from "@jest/globals";
import { parse } from "../../src/parser/parser";
import { TestCases as ListTestCases } from "../data/list.testcase";
import { TestCases as TextTestCases } from "../data/text.testcase";
import { TestCases as CodeBlockTestCases } from "../data/code_block.testcase";
import { tokenize } from "../../src/tokenizer/tokenizer";
import { ParseTreeNode } from "../../src/parser/parse_types";
describe("test text.testcase.ts", () => {
  it.each(TextTestCases)(
    'should parse "%p" correctly',
    ({ input, parse_expect }) => {
      const tokens = tokenize(input);
      const parseTree = parse(tokens);
      parse_expect(parseTree);
      expect(ParseTreeNode.toText(parseTree)).toEqual(input);
      expect(parseTree).toMatchSnapshot();
    },
  );
});

describe("test list.testcase.ts", () => {
  it.each(ListTestCases)(
    'should parse "%p" correctly',
    ({ input, parse_expect }) => {
      const tokens = tokenize(input);
      const parseTree = parse(tokens);
      parse_expect(parseTree);
      expect(ParseTreeNode.toText(parseTree)).toEqual(input);
      expect(parseTree).toMatchSnapshot();
    },
  );
});

describe("test code_block.testcase.ts", () => {
  it.each(CodeBlockTestCases)(
    'should parse "%p" correctly',
    ({ input, parse_expect }) => {
      const tokens = tokenize(input);
      const parseTree = parse(tokens);
      parse_expect(parseTree);
      expect(ParseTreeNode.toText(parseTree)).toEqual(input);
      expect(parseTree).toMatchSnapshot();
    },
  );
});
