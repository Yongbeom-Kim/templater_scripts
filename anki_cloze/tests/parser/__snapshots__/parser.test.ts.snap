// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test list.testcase.ts should parse "{"description": "Parse nested list of items", "input": "- Front - Back
  - Front = Back
1. Front - Back
  1. Front = Back
  2. Front = Back", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  ListNode {
    "children": [],
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 2,
            "lexeme": "Front",
            "line_pos": 1,
            "type": "text",
          },
          {
            "col_pos": 7,
            "lexeme": " ",
            "line_pos": 1,
            "type": "whitespace",
          },
          {
            "col_pos": 8,
            "lexeme": "-",
            "line_pos": 1,
            "type": "punctuation",
          },
          {
            "col_pos": 9,
            "lexeme": " ",
            "line_pos": 1,
            "type": "whitespace",
          },
          {
            "col_pos": 10,
            "lexeme": "Back",
            "line_pos": 1,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 14,
      "lexeme": "
",
      "line_pos": 1,
      "type": "newline",
    },
    "indent_level": 0,
    "marker": [
      {
        "col_pos": 0,
        "lexeme": "-",
        "line_pos": 1,
        "type": "punctuation",
      },
    ],
    "ordered": false,
    "type": "list",
  },
  ListNode {
    "children": [],
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 4,
            "lexeme": "Front",
            "line_pos": 2,
            "type": "text",
          },
          {
            "col_pos": 9,
            "lexeme": " ",
            "line_pos": 2,
            "type": "whitespace",
          },
          {
            "col_pos": 10,
            "lexeme": "=",
            "line_pos": 2,
            "type": "punctuation",
          },
          {
            "col_pos": 11,
            "lexeme": " ",
            "line_pos": 2,
            "type": "whitespace",
          },
          {
            "col_pos": 12,
            "lexeme": "Back",
            "line_pos": 2,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 16,
      "lexeme": "
",
      "line_pos": 2,
      "type": "newline",
    },
    "indent_level": 1,
    "marker": [
      {
        "col_pos": 2,
        "lexeme": "-",
        "line_pos": 2,
        "type": "punctuation",
      },
    ],
    "ordered": false,
    "type": "list",
  },
  ListNode {
    "children": [],
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 3,
            "lexeme": "Front",
            "line_pos": 3,
            "type": "text",
          },
          {
            "col_pos": 8,
            "lexeme": " ",
            "line_pos": 3,
            "type": "whitespace",
          },
          {
            "col_pos": 9,
            "lexeme": "-",
            "line_pos": 3,
            "type": "punctuation",
          },
          {
            "col_pos": 10,
            "lexeme": " ",
            "line_pos": 3,
            "type": "whitespace",
          },
          {
            "col_pos": 11,
            "lexeme": "Back",
            "line_pos": 3,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 15,
      "lexeme": "
",
      "line_pos": 3,
      "type": "newline",
    },
    "indent_level": 0,
    "marker": [
      {
        "col_pos": 0,
        "lexeme": "1",
        "line_pos": 3,
        "type": "number",
      },
      {
        "col_pos": 1,
        "lexeme": ".",
        "line_pos": 3,
        "type": "punctuation",
      },
    ],
    "ordered": true,
    "type": "list",
  },
  ListNode {
    "children": [],
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 5,
            "lexeme": "Front",
            "line_pos": 4,
            "type": "text",
          },
          {
            "col_pos": 10,
            "lexeme": " ",
            "line_pos": 4,
            "type": "whitespace",
          },
          {
            "col_pos": 11,
            "lexeme": "=",
            "line_pos": 4,
            "type": "punctuation",
          },
          {
            "col_pos": 12,
            "lexeme": " ",
            "line_pos": 4,
            "type": "whitespace",
          },
          {
            "col_pos": 13,
            "lexeme": "Back",
            "line_pos": 4,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 17,
      "lexeme": "
",
      "line_pos": 4,
      "type": "newline",
    },
    "indent_level": 1,
    "marker": [
      {
        "col_pos": 2,
        "lexeme": "1",
        "line_pos": 4,
        "type": "number",
      },
      {
        "col_pos": 3,
        "lexeme": ".",
        "line_pos": 4,
        "type": "punctuation",
      },
    ],
    "ordered": true,
    "type": "list",
  },
  ListNode {
    "children": [],
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 5,
            "lexeme": "Front",
            "line_pos": 5,
            "type": "text",
          },
          {
            "col_pos": 10,
            "lexeme": " ",
            "line_pos": 5,
            "type": "whitespace",
          },
          {
            "col_pos": 11,
            "lexeme": "=",
            "line_pos": 5,
            "type": "punctuation",
          },
          {
            "col_pos": 12,
            "lexeme": " ",
            "line_pos": 5,
            "type": "whitespace",
          },
          {
            "col_pos": 13,
            "lexeme": "Back",
            "line_pos": 5,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": undefined,
    "indent_level": 1,
    "marker": [
      {
        "col_pos": 2,
        "lexeme": "2",
        "line_pos": 5,
        "type": "number",
      },
      {
        "col_pos": 3,
        "lexeme": ".",
        "line_pos": 5,
        "type": "punctuation",
      },
    ],
    "ordered": true,
    "type": "list",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Mixed whitespace and newlines", "input": " 	 
 	 ", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 3,
      "lexeme": "
",
      "line_pos": 1,
      "type": "newline",
    },
    "indent_level": 2,
    "type": "text",
  },
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [],
        "type": "text",
      },
    ],
    "endingNewline": undefined,
    "indent_level": 2,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Multiple newlines", "input": "

", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 0,
      "lexeme": "

",
      "line_pos": 1,
      "type": "newline",
    },
    "indent_level": 0,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Simple text with newline", "input": "Hello 123
World", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "Hello",
            "line_pos": 1,
            "type": "text",
          },
          {
            "col_pos": 5,
            "lexeme": " ",
            "line_pos": 1,
            "type": "whitespace",
          },
          {
            "col_pos": 6,
            "lexeme": "123",
            "line_pos": 1,
            "type": "number",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 9,
      "lexeme": "
",
      "line_pos": 1,
      "type": "newline",
    },
    "indent_level": 0,
    "type": "text",
  },
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "World",
            "line_pos": 2,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": undefined,
    "indent_level": 0,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Simple text with numbers", "input": "123 456", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "123",
            "line_pos": 1,
            "type": "number",
          },
          {
            "col_pos": 3,
            "lexeme": " ",
            "line_pos": 1,
            "type": "whitespace",
          },
          {
            "col_pos": 4,
            "lexeme": "456",
            "line_pos": 1,
            "type": "number",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": undefined,
    "indent_level": 0,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Text with mixed whitespace and newlines", "input": "Hello 123
 	World
456", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "Hello",
            "line_pos": 1,
            "type": "text",
          },
          {
            "col_pos": 5,
            "lexeme": " ",
            "line_pos": 1,
            "type": "whitespace",
          },
          {
            "col_pos": 6,
            "lexeme": "123",
            "line_pos": 1,
            "type": "number",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 9,
      "lexeme": "
",
      "line_pos": 1,
      "type": "newline",
    },
    "indent_level": 0,
    "type": "text",
  },
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 2,
            "lexeme": "World",
            "line_pos": 2,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 7,
      "lexeme": "
",
      "line_pos": 2,
      "type": "newline",
    },
    "indent_level": 2,
    "type": "text",
  },
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "456",
            "line_pos": 3,
            "type": "number",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": undefined,
    "indent_level": 0,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Text with multiple newlines", "input": "Multiple


Newlines


", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "Multiple",
            "line_pos": 1,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 8,
      "lexeme": "


",
      "line_pos": 1,
      "type": "newline",
    },
    "indent_level": 0,
    "type": "text",
  },
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [
          {
            "col_pos": 0,
            "lexeme": "Newlines",
            "line_pos": 4,
            "type": "text",
          },
        ],
        "type": "text",
      },
    ],
    "endingNewline": {
      "col_pos": 8,
      "lexeme": "


",
      "line_pos": 4,
      "type": "newline",
    },
    "indent_level": 0,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should parse "{"description": "Whitespace characters", "input": " 	 ", "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  TextLineNode {
    "contents": [
      TextNode {
        "contents": [],
        "type": "text",
      },
    ],
    "endingNewline": undefined,
    "indent_level": 2,
    "type": "text",
  },
]
`;
